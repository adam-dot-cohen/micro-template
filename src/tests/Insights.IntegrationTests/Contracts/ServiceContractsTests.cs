using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using Google.Protobuf.Reflection;
using Grpc.Core;
using Laso.Insights.IntegrationTests.Extensions;
using Laso.IntegrationEvents;
using Xunit;

namespace Laso.Insights.IntegrationTests.Contracts
{
    public class ServiceContractsTests
    {
        private static readonly ICollection<Type> AllReferencedTypes = Directory.GetFiles(AppDomain.CurrentDomain.BaseDirectory, "*.dll")
            .Select(AssemblyName.GetAssemblyName)
            .Where(x => x.Name?.StartsWith("Laso") == true)
            .Select(Assembly.Load)
            .SelectMany(x => x.GetTypes())
            .ToList();

        [Fact]
        public void Proto_files_should_not_have_breaking_changes()
        {
            static string GetType(FieldDescriptor field)
            {
                return field.FieldType switch
                {
                    FieldType.Message => field.MessageType.FullName,
                    FieldType.Enum => field.EnumType.FullName,
                    FieldType.Group => throw new NotSupportedException(),
                    _ => field.FieldType.ToString()
                };
            }

            AllReferencedTypes
                .Select(x => x.GetCustomAttribute<BindServiceMethodAttribute>(false)?.BindType)
                .Where(x => x != null)
                .Select(x => ((ServiceDescriptor) x.GetProperty("Descriptor").GetValue(null)).File)
                .Distinct(x => x.Name)
                .ToList()
                .To(x => new
                {
                    Methods = x
                        .SelectMany(y => y.Services, (y, z) => new { Proto = y.Name, Service = z })
                        .SelectMany(y => y.Service.Methods, (y, z) => new
                        {
                            y.Proto,
                            Service = y.Service.FullName,
                            z.Name,
                            InputType = z.InputType.FullName,
                            OutputType = z.OutputType.FullName
                        }),
                    Fields = x
                        .SelectMany(y => y.MessageTypes, (y, z) => new { Proto = y.Name, MessageType = z })
                        .SelectMany(y => y.MessageType.Fields.InFieldNumberOrder(), (y, z) => new
                        {
                            y.Proto,
                            MessageType = y.MessageType.FullName,
                            z.Name,
                            Type = GetType(z),
                            z.FieldNumber,
                            z.IsRequired,
                            z.IsRepeated,
                            //z.IsPacked, This seems to be bugged
                            z.IsMap,
                            ContainingOneof = z.ContainingOneof?.Name
                        }),
                    EnumValues = x
                        .SelectMany(y => y.EnumTypes, (y, z) => new { Proto = y.Name, EnumType = z })
                        .SelectMany(y => y.EnumType.Values, (y, z) => new
                        {
                            y.Proto,
                            EnumType = y.EnumType.FullName,
                            z.Name,
                            z.Number
                        })
                })
                .ApproveIf(x => x.Approved.Methods.All(y => x.Received.Methods.Contains(y))
                                && x.Approved.Fields.All(y => x.Received.Fields.Contains(y))
                                && x.Approved.EnumValues.All(y => x.Received.EnumValues.Contains(y))
                                && x.Received.EnumValues.All(y => x.Approved.EnumValues.Contains(y)));
        }

        [Fact]
        public void Integration_events_should_not_have_breaking_changes()
        {
            static void VisitMessageTypes(Type type, ICollection<Type> messageTypes, ICollection<Type> enumTypes)
            {
                if (messageTypes.Contains(type))
                    return;

                messageTypes.Add(type);

                type.GetProperties()
                    .Select(x => x.PropertyType.GetNonNullableType())
                    .ForEach(x =>
                    {
                        //Add more "primitive" types here
                        if (x.IsPrimitive || x == typeof(string) || x == typeof(DateTime) || x == typeof(Guid) || x == typeof(DateTimeOffset))
                            return;

                        if (x.IsEnum)
                            enumTypes.Add(x);

                        VisitMessageTypes(x.Closes(typeof(IEnumerable<>), out var args) ? args.First()[0] : x, messageTypes, enumTypes);
                    });
            }

            AllReferencedTypes
                .Where(x => typeof(IIntegrationEvent).IsAssignableFrom(x) && !x.IsInterface && !x.IsAbstract)
                .ToList()
                .To(x =>
                {
                    var messageTypes = new HashSet<Type>();
                    var enumTypes = new HashSet<Type>();

                    x.ForEach(y => VisitMessageTypes(y, messageTypes, enumTypes));

                    return new
                    {
                        Fields = messageTypes.SelectMany(y => y.GetProperties(), (y, z) => new
                        {
                            MessageType = y.FullName,
                            z.Name,
                            Type = GetTypeName(z.PropertyType)
                        }),
                        EnumValues = enumTypes.SelectMany(y => Enum.GetValues(y).Cast<Enum>(), (y, z) => new
                        {
                            EnumType = y.FullName,
                            Name = z.ToString(),
                            Value = z.GetValue()
                        })
                    };
                })
                .ApproveIf(x => x.Approved.Fields.All(y => x.Received.Fields.Contains(y))
                                && x.Approved.EnumValues.All(y => x.Received.EnumValues.Contains(y))
                                && x.Received.EnumValues.All(y => x.Approved.EnumValues.Contains(y)));
        }

        private static string GetTypeName(Type type)
        {
            if (type != typeof(string) && type.Closes(typeof(IEnumerable<>), out var args))
               return $"List<{args.First()[0].FullName}>";

            return type.FullName;
        }
    }
}
