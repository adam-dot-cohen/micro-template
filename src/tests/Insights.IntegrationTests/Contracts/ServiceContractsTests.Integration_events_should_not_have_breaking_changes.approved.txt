{
  "Fields": [
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.PartnerFilesReceivedEvent",
      "Name": "FileBatchId",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.PartnerFilesReceivedEvent",
      "Name": "Timestamp",
      "Type": "System.DateTimeOffset"
    },
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.PartnerFilesReceivedEvent",
      "Name": "CorrelationId",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.PartnerFilesReceivedEvent",
      "Name": "PartnerId",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.PartnerFilesReceivedEvent",
      "Name": "PartnerName",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.PartnerFilesReceivedEvent",
      "Name": "Files",
      "Type": "List<Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.BlobFileInfo>"
    },
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.BlobFileInfo",
      "Name": "Id",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.BlobFileInfo",
      "Name": "Uri",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.BlobFileInfo",
      "Name": "ContentType",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.BlobFileInfo",
      "Name": "ContentLength",
      "Type": "System.Int64"
    },
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.BlobFileInfo",
      "Name": "ETag",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.BlobFileInfo",
      "Name": "DataCategory",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.BlobFileInfo",
      "Name": "EffectiveDate",
      "Type": "System.DateTimeOffset"
    },
    {
      "MessageType": "Laso.AdminPortal.Infrastructure.Monitoring.DataQualityPipeline.IntegrationEvents.BlobFileInfo",
      "Name": "TransmissionTime",
      "Type": "System.DateTimeOffset"
    },
    {
      "MessageType": "Laso.Identity.Core.IntegrationEvents.PartnerCreatedEventV1",
      "Name": "Id",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Identity.Core.IntegrationEvents.PartnerCreatedEventV1",
      "Name": "Name",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Identity.Core.IntegrationEvents.PartnerCreatedEventV1",
      "Name": "NormalizedName",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.FTPCredentialsCreationFailedEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.FTPCredentialsCreationFailedEvent",
      "Name": "PartnerId",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.FTPCredentialsCreationFailedEvent",
      "Name": "Reason",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.FTPCredentialsCreatedEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.FTPCredentialsCreatedEvent",
      "Name": "PartnerId",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.FTPCredentialsCreatedEvent",
      "Name": "UsernameSecret",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.FTPCredentialsCreatedEvent",
      "Name": "PasswordSecret",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.PartnerAccountCreatedEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.PartnerAccountCreatedEvent",
      "Name": "PartnerId",
      "Type": "System.Guid"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.PartnerAccountCreationFailedEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.PartnerAccountCreationFailedEvent",
      "Name": "PartnerId",
      "Type": "System.Guid"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.PartnerAccountCreationFailedEvent",
      "Name": "ErrorMessage",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.PartnerAccountDeletedEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.PartnerAccountDeletedEvent",
      "Name": "PartnerId",
      "Type": "System.Guid"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.DeletePartnerAccountFailedEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.DeletePartnerAccountFailedEvent",
      "Name": "PartnerId",
      "Type": "System.Guid"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.DeletePartnerAccountFailedEvent",
      "Name": "ErrorMessage",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.RotateAdminPasswordFailedEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.RotateAdminPasswordFailedEvent",
      "Name": "Reason",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.RotatedAdminPasswordEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.RotatedAdminPasswordEvent",
      "Name": "SecretUsedName",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.RotatedAdminPasswordEvent",
      "Name": "ToVersion",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.UpdatedPartnerPasswordEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.UpdatedPartnerPasswordEvent",
      "Name": "PartnerId",
      "Type": "System.Guid"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.FailedToUpdatePartnerPasswordEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.FailedToUpdatePartnerPasswordEvent",
      "Name": "PartnerId",
      "Type": "System.Guid"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.SFTP.FailedToUpdatePartnerPasswordEvent",
      "Name": "ErrorMessage",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.Encryption.PartnerPgpKeySetCreatedEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.Encryption.PartnerPgpKeySetCreatedEvent",
      "Name": "PartnerId",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.Encryption.PartnerPgpKeySetCreationFailedEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.Encryption.PartnerPgpKeySetCreationFailedEvent",
      "Name": "PartnerId",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.Encryption.PartnerPgpKeySetCreationFailedEvent",
      "Name": "Reason",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.AzureResources.PartnerColdStorageCreatedEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.AzureResources.PartnerColdStorageCreatedEvent",
      "Name": "PartnerId",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.AzureResources.PartnerColdStorageCreationFailedEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.AzureResources.PartnerColdStorageCreationFailedEvent",
      "Name": "PartnerId",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.AzureResources.PartnerColdStorageCreationFailedEvent",
      "Name": "Reason",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.AzureResources.PartnerDataProcessingDirCreatedEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.AzureResources.PartnerDataProcessingDirCreatedEvent",
      "Name": "PartnerId",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.AzureResources.PartnerDataProcessingDirFailedEvent",
      "Name": "OnUtc",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.AzureResources.PartnerDataProcessingDirFailedEvent",
      "Name": "PartnerId",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.AzureResources.PartnerDataProcessingDirFailedEvent",
      "Name": "Reason",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.AzureResources.EscrowPartnerStorageCreatedEvent",
      "Name": "PartnerId",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.AzureResources.EscrowPartnerStorageCreationFailedEvent",
      "Name": "PartnerId",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.Messaging.AzureResources.EscrowPartnerStorageCreationFailedEvent",
      "Name": "Reason",
      "Type": "System.String"
    },
    {
      "MessageType": "Laso.Provisioning.Core.IntegrationEvents.ProvisioningCompletedEvent",
      "Name": "CompletedOn",
      "Type": "System.DateTime"
    },
    {
      "MessageType": "Laso.Provisioning.Core.IntegrationEvents.ProvisioningCompletedEvent",
      "Name": "PartnerId",
      "Type": "System.String"
    }
  ],
  "EnumValues": []
}